[quant][fx][pt2e] Refactor prepare so it's aligned better with the new API plan in pt2e

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94011

Summary:
There are three things that happens in the current prepare code,
(1). user express their intention of how they want the model to be quantized with QConfigMapping, we translate that to
node.meta["target_dtype_info"]
(2). we validate the setting against BackendConfig
(3). insert observers based on the validated node.meta["target_dtype_info"]

previously (2) and (3) are mixed together, this PR tries to move (2) closer to (1), with one edge case left, this refactor
moves us closer to our target design for quantization in pytorch 2.0 export path

this is a follow up PR for https://github.com/pytorch/pytorch/pull/92641

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][fx][pt2e] Refactor prepare so it's aligned better with the new API plan in pt2e

Summary:
There are three things that happens in the current prepare code,
(1). user express their intention of how they want the model to be quantized with QConfigMapping, we translate that to
node.meta["target_dtype_info"]
(2). we validate the setting against BackendConfig
(3). insert observers based on the validated node.meta["target_dtype_info"]

previously (2) and (3) are mixed together, this PR tries to move (2) closer to (1), with one edge case left, this refactor
moves us closer to our target design for quantization in pytorch 2.0 export path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][pt2e] Refactor prepare so it's aligned better with the new API plan in pt2e"


Summary:
There are three things that happens in the current prepare code,
(1). user express their intention of how they want the model to be quantized with QConfigMapping, we translate that to
node.meta["target_dtype_info"]
(2). we validate the setting against BackendConfig
(3). insert observers based on the validated node.meta["target_dtype_info"]

previously (2) and (3) are mixed together, this PR tries to move (2) closer to (1), with one edge case left, this refactor
moves us closer to our target design for quantization in pytorch 2.0 export path

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1205,10 +1208,14 @@ def insert_observers_for_model(
     # initalize target_dtype_info
-    DEFAULT_QCONFIG_FOR_TARGET_DTYPE_INFO = \
-        QConfig(activation=PlaceholderObserver.with_args(dtype=torch.float32), weight=PlaceholderObserver.with_args(dtype=torch.float32))
+    DEFAULT_QCONFIG_FOR_TARGET_DTYPE_INFO = (
```
should this live on backendconfig instead of here?  Seems like this is saying "configuration for default dtype", and since it's default that points to backendconfig instead of qconfig_mapping?

jerryzh168 replied:
I think this can live in the new "QTagger" API code, this is just temporary

vkuzo replied:
I think we chatted before that the programmable API is for user intent.  A default output dtype config doesn't seem like user intent, but rather a default capability which makes sense in a target environment.

So, sounds like it would explicitly not be in something like `QTaggerFutureNameTBD`, and instead be in BackendConfig.  What do you think?

jerryzh168 replied:
> A default output dtype config doesn't seem like user intent

I feel it should be, this is like user doing QConfigMapping.set_global(None)?

vkuzo replied:
I think we should avoid using `QTaggerFutureNameTBD` as a catch-all for "all configuration things".  It may be the API exposed to users to programmably adjust intent, but it might be clearer to chat about the underlying components when possible.

For "default dtype", it seems that the value has to depend on the backend.  For example, for common NVIDIA gpu/x86 cpu and PyTorch in general, its fp32.  For some custom DSP that does not support fp32, it may be fp16, or even an integer type.

I think when the user is saying `foo.bar.qconfig=None`, the user intent is "use whatever is default for target environment".  The "what is default in target environment" piece is coming from the backend, not the user.

vkuzo replied:
I'd push for `QTaggerFutureNameTBD` to not have any configs or defaults at all in its source code - those should be clearly encapsulated as separate objects and imported.

jerryzh168 replied:
I feel the default is a part of the original unquantized pytorch model? let's say we are quantizing for a DSP that only supports int8, if we start with a floating point pytorch model, can we quantize the model to int8 when user just says QConfigMapping().set_global(None)?

vkuzo replied:
> I feel the default is don't quantize the model and the default dtype should just be derived from the original unquantized model?

that makes sense as as well.

I think the confusing thing in the current code is that the default is created inside of a utility function.  Seems like this should either be a configuration (default means default of target backend), or it should read the existing dtype (default means what the input program is doing).  Neither of those things should care about `prepare.py`.

jerryzh168 replied:
I think the current state is that it is trying to do the latter, we can certainly improvement this by either just use "None" or read from node.meta["val"] in the future, this PR is just a refactor so I hope to do these fixes in the future

vkuzo replied:
I think doing this here is pretty confusing to non-experts as it changes the role of QConfig from "comes from user or default" to "can be created inside code", can we add a comment on the future plan and fix it soon?

vkuzo replied:
or how about just

```
// qconfig.py
default_fp32_placeholder_qconfig = ...

// this file can just reference that object
```

jerryzh168 replied:
sure, can I add a private one: `_default_fp32_placeholder_qconfig`? we may delete this soon

jerryzh168 replied:
@vkuzo updated, please take a look again

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -420,7 +491,7 @@ def _get_target_activation_dtype_for_node(
 
     elif node.op == 'output':
-        if outputs_seen_counter in output_quantized_idxs:
+        if output_node_to_output_index[node] in output_quantized_idxs:
```
can we add a comment that creating placeholderobserver below is temporary

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1129,38 +1200,136 @@ def insert_observers_for_model(
+        )
+    )
+    for node in model.graph.nodes:
+        node.meta["target_dtype_info"] = {
```
does this assume one input/one output?  Thoughts on what happens when there are multiple inputs / output is not a node?

jerryzh168 replied:
currently this means we'll use these observers/fqs for all the input and outputs that need to be quantized

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1129,38 +1200,136 @@ def insert_observers_for_model(
+    inputs_seen_counter = 0
+    outputs_seen_counter = 0
+    placeholder_node_to_input_index: Dict[Node, int] = {}
+    # TODO: we probably don't need this counter since each graph will only have
```
the original plan was to implement anything a user could do, such as 

```
UserModule(foo, (bar, baz), {'foo': bar}, ...)
```

or 

```
return 1, (2, 3), {4: {5: 6}}
```

all or that was supported by PyTorch and FX symbolic trace, not sure if the story is different in PT2.0 export?

jerryzh168 replied:
my understanding is that for input we will have multiple placeholder nodes, but for output we only have one node
```
return 1, (2, 3), {4: {5: 6}}
```
will be one output node with an argument of a tuple(1, (2, 3), {4: {5: 6}}), also cc @qihqi to confirm as well

vkuzo replied:
yeah.  The counter is for the individual parts of the outputs, it shouldn't matter that it's a single node at the end.  For example, if it's `return (a, b)`, the user should be able to configure the dtype of `b`.

jerryzh168 replied:
yeah agree, we just don't have code to handle this right now

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1129,38 +1200,136 @@ def insert_observers_for_model(
 
-    # Second, for nodes with known input dtypes, propagate them throughout the
+    # Step 1, set the observer or fake quantize module constructor for each node in the
+    # matched_node_pattern
```
for my own understanding, should this be "each node in pattern", or "subgraph endpoints"?  I thought it was subgraph endpoints

jerryzh168 replied:
right now it is still each node in pattern, in the future we'll change this to "subgraph endpoints"

jerryzh168 committed:
Update on "[quant][fx][pt2e] Refactor prepare so it's aligned better with the new API plan in pt2e"


Summary:
There are three things that happens in the current prepare code,
(1). user express their intention of how they want the model to be quantized with QConfigMapping, we translate that to
node.meta["target_dtype_info"]
(2). we validate the setting against BackendConfig
(3). insert observers based on the validated node.meta["target_dtype_info"]

previously (2) and (3) are mixed together, this PR tries to move (2) closer to (1), with one edge case left, this refactor
moves us closer to our target design for quantization in pytorch 2.0 export path

this is a follow up PR for https://github.com/pytorch/pytorch/pull/92641

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1129,38 +1202,136 @@ def insert_observers_for_model(
         if node.op == "placeholder":
+            placeholder_node_to_input_index[node] = inputs_seen_counter
             inputs_seen_counter += 1
         if node.op == "output":
+            output_node_to_output_index[node] = outputs_seen_counter
             outputs_seen_counter += 1
```
@vkuzo the comment is saying, we are checking "output" node here, and we'll always only have one output node

vkuzo replied:
yeah, the code is probably wrong.  What this should be doing is looking at the things the user wants to return.

jerryzh168 replied:
yeah, we can fix this in the new API

jerryzh168 committed:
Update on "[quant][fx][pt2e] Refactor prepare so it's aligned better with the new API plan in pt2e"


Summary:
There are three things that happens in the current prepare code,
(1). user express their intention of how they want the model to be quantized with QConfigMapping, we translate that to
node.meta["target_dtype_info"]
(2). we validate the setting against BackendConfig
(3). insert observers based on the validated node.meta["target_dtype_info"]

previously (2) and (3) are mixed together, this PR tries to move (2) closer to (1), with one edge case left, this refactor
moves us closer to our target design for quantization in pytorch 2.0 export path

this is a follow up PR for https://github.com/pytorch/pytorch/pull/92641

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][pt2e] Refactor prepare so it's aligned better with the new API plan in pt2e"


Summary:
There are three things that happens in the current prepare code,
(1). user express their intention of how they want the model to be quantized with QConfigMapping, we translate that to
node.meta["target_dtype_info"]
(2). we validate the setting against BackendConfig
(3). insert observers based on the validated node.meta["target_dtype_info"]

previously (2) and (3) are mixed together, this PR tries to move (2) closer to (1), with one edge case left, this refactor
moves us closer to our target design for quantization in pytorch 2.0 export path

this is a follow up PR for https://github.com/pytorch/pytorch/pull/92641

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][pt2e] Refactor prepare so it's aligned better with the new API plan in pt2e"


Summary:
There are three things that happens in the current prepare code,
(1). user express their intention of how they want the model to be quantized with QConfigMapping, we translate that to
node.meta["target_dtype_info"]
(2). we validate the setting against BackendConfig
(3). insert observers based on the validated node.meta["target_dtype_info"]

previously (2) and (3) are mixed together, this PR tries to move (2) closer to (1), with one edge case left, this refactor
moves us closer to our target design for quantization in pytorch 2.0 export path

this is a follow up PR for https://github.com/pytorch/pytorch/pull/92641

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -420,7 +490,9 @@ def _get_target_activation_dtype_for_node(
     elif node.op == 'output':
-        if outputs_seen_counter in output_quantized_idxs:
+        # Note: creating placeholder observer here is temporary, it will be moved
+        # to the new programmable API when that is ready
```
I think I'm still confused on the "programmable" part fo the programmable API.  I thought the "programmable" part would be enabling the user to do something they cannot do with simple defaults because their modeling code is complicated.  Not sure why this specific case would be related.  I think the user should be able to override this in a programmable way, but the default shouldn't have to go through the programmable way.

vkuzo replied:
for L497:498, can that read from qconfig.py too instead of creating new hardcoded observer objects?

jerryzh168 replied:
this comment is saying we should not be creating new placeholder observer configs in prepare, this is just temporary

sure I can move L497:498 to qconfig.py as well

jerryzh168 replied:
> I think the user should be able to override this in a programmable way, but the default shouldn't have to go through the programmable way.

not sure what this means, in the end both programmable api or config api should end up with configuring the same thing, which is something like:
node.meta["input_act_obs_or_fq_ctr"] = ...
node.meta["output_act_obs_or_fq_ctr"] = ...
I think.

vkuzo replied:
how about something like:
* a reasonable default is specified in a configuration
* user can override things in a programmable way

Both of those things can live under the `QTaggerFutureNameTBD`, but in different places of it.

jerryzh168 replied:
yeah that is the plan, let's discuss in details when I put up that PR, current PR is still walking towards the goal of separating the configuration related things to a separate class

jerryzh168 committed:
Update on "[quant][fx][pt2e] Refactor prepare so it's aligned better with the new API plan in pt2e"


Summary:
There are three things that happens in the current prepare code,
(1). user express their intention of how they want the model to be quantized with QConfigMapping, we translate that to
node.meta["target_dtype_info"]
(2). we validate the setting against BackendConfig
(3). insert observers based on the validated node.meta["target_dtype_info"]

previously (2) and (3) are mixed together, this PR tries to move (2) closer to (1), with one edge case left, this refactor
moves us closer to our target design for quantization in pytorch 2.0 export path

this is a follow up PR for https://github.com/pytorch/pytorch/pull/92641

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge -g

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge -f "in progress test is running for 12 hours and all other tests are passing" 

(https://github.com/pytorch/pytorch/actions/runs/4107390820/jobs/7087237833)

jerryzh168 commented:
@pytorchbot merge -f "in progress test is running for 12 hours and all other tests are passing"
